### PR TITLE
Support cf.switch in CoreDSL

### PR DESCRIPTION
**Pull request by PetrifiedPanda**
Implemented all necessary steps to support cf.switch in shortnail. This consists of a pass that converts control flow operations back to scf and a pass that converts the generated index_switch operations to scf.if operations to avoid using the index type.
Also used dependentDialects for CoreDSL, so all other needed dialects are loaded automatically when adding the CoreDSL dialect

### CODE REVIEW

File: lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+struct CoreDSLSwitchToIf
+    : public mlir::shortnail::impl::CoreDSLSwitchToIfBase<CoreDSLSwitchToIf> {
+  using CoreDSLSwitchToIfBase::CoreDSLSwitchToIfBase;
+
+  void runOnOperation() override {
+    auto isaxOp = getOperation();
+    MLIRContext &ctx = getContext();
+    // First convert the cf operations
+    OpPassManager cfToSCFPM{isaxOp.getOperationName()};
+    cfToSCFPM.addPass(createCoreDSLLiftCFToSCF());
```
Comment (7FM) [inline lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp]: Is there a particular reason why this is split into two passes?

Comment (PetrifiedPanda) [reply lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp]: The CoreDSLToPy pass only uses the conversion to SCF and then converts that to python. It could probably be adapted to operate on the result of the switch to if pass as well. Then everything could be moved into a single pass.

Comment (PetrifiedPanda) [reply lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp]: This would also probably be good, because then no special handling for scf.index_switch would be needed in CoreDSLToPy anymore

Comment (7FM) [reply lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp]: Yes, please do this :)

Comment (7FM) [reply lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp]: And don't explicitly call shortnail passes within other shortnail passes, you can expect that users / nailgun will call those passes sequentially

Comment (PetrifiedPanda) [reply lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp]: Will keep that in mind. Fix is pushed

File: lib/Conversion/CoreDSLToPy/CoreDSLToPy.cpp [hunk 2]
```diff
@@ -828,8 +851,10 @@ struct CoreDSLToPy
   void runOnOperation() override {
     auto isaxOp = getOperation();
 
-    // Inline all functions into the InstructionOps
     OpPassManager inlinePM(isaxOp.getOperationName());
+    // Before inliing, convert all cf operations to scf
+    inlinePM.addPass(createCoreDSLSwitchToIf());
```
Comment (7FM) [inline lib/Conversion/CoreDSLToPy/CoreDSLToPy.cpp]: same here, expect nailgun to run this step

File: lib/Conversion/CoreDSLToPy/CoreDSLToPy.cpp [hunk 3]
```diff
@@ -575,6 +575,29 @@ static WalkResult emitArithOp(mlir::raw_indented_ostream &os, Operation *op) {
         os << ")\n";
         return WalkResult::advance();
       })
+      // NOTE: The following two ops are generated by ControlFlowToSCF
```
Comment (7FM) [inline lib/Conversion/CoreDSLToPy/CoreDSLToPy.cpp]: is this still the case after createCoreDSLSwitchToIf?

File: lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    // index_cast of the argument of the original cf.switch
+    assert(isa<arith::IndexCastUIOp>(arg.getDefiningOp()) &&
+           "ControlFlowToSCF pass should only generate arith::IndexCastUIOp");
+    auto indexCast = dyn_cast<arith::IndexCastUIOp>(arg.getDefiningOp());
+    auto nonIndexArg = indexCast.getOperand();
+    auto cmpType = dyn_cast<IntegerType>(nonIndexArg.getType());
+    assert(cmpType);
+    auto caseAttr = IntegerAttr::get(cmpType, caseVal);
+    // Because cf.switch only supports signless integers as arguments, we can't
+    // use hwarith operations for the comparisons
```
Comment (7FM) [inline lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp]: nonIndexArg should be a hwarith typed value (or one hwarith.cast away from one), right?
Can't we use this to directly emit hwarith operations instead of taking the detour via the arith dialect?

File: lib/Conversion/CMakeLists.txt [hunk 5]
```diff
@@ -1,3 +1,4 @@
 add_subdirectory(AnalyzeISAX)
 add_subdirectory(CoreDSLToPy)
+add_subdirectory(CoreDSLSwitchToIf)
```
Comment (7FM) [inline lib/Conversion/CMakeLists.txt]: keep the alphabetical order pls :)

File: tools/shortnail-opt/CMakeLists.txt [hunk 6]
```diff
@@ -27,6 +27,7 @@ target_link_libraries(shortnail-opt
     ShortnailCoreDSL
     ShortnailAnalyzeISAX
     ShortnailCoreDSLToPy
+    ShortnailCoreDSLSwitchToIf
```
Comment (7FM) [inline tools/shortnail-opt/CMakeLists.txt]: same here. but seems like I messed up myself with the analyzeISAX pass, fix that too pls

File: lib/Conversion/CoreDSLToPy/CoreDSLToPy.cpp (hunk 3 above)
Comment (PetrifiedPanda) [reply lib/Conversion/CoreDSLToPy/CoreDSLToPy.cpp]: Yes. Though I just realized that hw.constant was never generated before I added the conversion to if back. But with the if conversion it is generated. It crashes on the getAPSInt call however, so it may have never been correct in the first place. I think the easiest way to fix this is to just remove the arith operations entirely like in the other comment

File: lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp (hunk 4 above)
Comment (PetrifiedPanda) [reply lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp]: Good point. I'll work on this tomorrow.

Comment (PetrifiedPanda) [reply lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp]: Fixed: https://github.com/esa-tu-darmstadt/Shortnail/pull/10/changes/692fc4c09c1bc1effa38cba166caac60c05b39d9

File: lib/Conversion/CoreDSLToPy/CoreDSLToPy.cpp (hunk 3 above)
Comment (PetrifiedPanda) [reply lib/Conversion/CoreDSLToPy/CoreDSLToPy.cpp]: As expected, these are no longer needed after only generating hwarith ops: https://github.com/esa-tu-darmstadt/Shortnail/pull/10/changes/0661af694fc6927fca7201220617c02be7384ba6

File: lib/Conversion/CoreDSLToPy/CoreDSLToPy.cpp (hunk 2 above)
Comment (PetrifiedPanda) [reply lib/Conversion/CoreDSLToPy/CoreDSLToPy.cpp]: Fixed: https://github.com/esa-tu-darmstadt/Shortnail/pull/10/changes/0e32a8250f565a43e8e121aabbe1f854d6b04e59

File: lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    // Non index arg must be a signless integer, but it has to have been created
+    // by converting a signed integer to signless
+    assert(isa<hwarith::CastOp>(nonIndexArg.getDefiningOp()));
+    auto signlessCast = cast<hwarith::CastOp>(nonIndexArg.getDefiningOp());
+    auto hwarithValue = signlessCast.getOperand();
+    auto cmpType = dyn_cast<IntegerType>(hwarithValue.getType());
+    assert(cmpType);
+    auto caseAttr = IntegerAttr::get(cmpType, caseVal);
+    auto constant =
+        hwarith::ConstantOp::create(rewriter, loc, cmpType, caseAttr);
```
Comment (7FM) [inline lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp]: nit: inline caseAttr, rm assert -> use cast<> instead

File: tools/shortnail-opt/CMakeLists.txt (hunk 6 above)
Comment (PetrifiedPanda) [reply tools/shortnail-opt/CMakeLists.txt]: Fixed https://github.com/esa-tu-darmstadt/Shortnail/pull/10/changes/d65d417cf7d5fd9a66ac76193d86823dd9090d7d

File: lib/Conversion/CMakeLists.txt (hunk 5 above)
Comment (PetrifiedPanda) [reply lib/Conversion/CMakeLists.txt]: Fixed https://github.com/esa-tu-darmstadt/Shortnail/pull/10/changes/bb9a9b1a8566246164680b1bbd7a583783c2777a

File: lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp (hunk 7 above)
Comment (PetrifiedPanda) [reply lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp]: So `auto constant = hwarith::ConstantOp::create(rewriter, loc, cmpType, IntegerAttr::get(cmpType, caseVal));` and`auto cmpType = cast<IntegerType>(hwarithValue.getType())`? Sorry, I don't know how to suggest changes on Github.

Comment (7FM) [reply lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp]: exactly :)

Comment (PetrifiedPanda) [reply lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp]: Done https://github.com/esa-tu-darmstadt/Shortnail/pull/10/changes/1db38d30b46deb5bc215c6a6de7e1788ec9c1b99

**Comment by 7FM** [comment]
why does removeDeadValues create ub.poison and why do we need this pass in the first place?

**Comment by PetrifiedPanda** [comment]
I used the removeDeadValues pass, because every time I tried to delete the casts, the rewrite crashed with the error "dialect conversion attempted to replace a root operation that has no parent block".
The pass seems to first insert poison for all values then delete them. But the error only started when I switched to using hwarith. Maybe because it now has to delete two operations (cast to index and then cast to signless)

File: lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+      markAllAnalysesPreserved();
+      // If nothing changed, there is no need to convert scf.index_switch to
+      // scf.if, as we expect there to not be any index switch ops other than
+      // the one generated by converting cf.switch ops
+      return;
+    }
+    MLIRContext &ctx = getContext();
+    // First convert the cf operations
+    OpPassManager cfToSCFPM{isaxOp.getOperationName()};
+    if (failed(runPipeline(cfToSCFPM, isaxOp))) {
```
Comment (7FM) [inline lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp]: empty pass run

File: lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+    target.addIllegalOp<scf::IndexSwitchOp>();
+    patterns.insert<IndexSwitchToSCFIf>(&ctx);
+    if (failed(applyFullConversion(isaxOp, target, std::move(patterns)))) {
+      return signalPassFailure();
+    }
+    // Run a dead value removal pass, as the index casts are now dead
+    // NOTE: workaround to prevent crashes when deleting the index cast in
+    // IndexSwitchToSCFIf::run
+    OpPassManager finalPass{isaxOp.getOperationName()};
+    finalPass.addPass(createRemoveDeadValuesPass());
```
Comment (7FM) [inline lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp]: have you tried smth like:
```
  isaxOp->walk([](arith::IndexCastUIOp op) {
    if (op->use_empty())
      op->erase();
  });
```
Instead?

File: lib/Conversion/CoreDSLLegalizeCF/CoreDSLLegalizeCF.cpp [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+    }
+    res =
+        isaxOp->walk<ORDER>(RegionToSCFConverter<func::FuncOp>{*this, changed});
+    if (res == WalkResult::interrupt()) {
+      return signalPassFailure();
+    }
+    res = isaxOp->walk<ORDER>(
+        RegionToSCFConverter<scf::ExecuteRegionOp>{*this, changed});
+    if (res == WalkResult::interrupt()) {
+      return signalPassFailure();
```
Comment (7FM) [inline lib/Conversion/CoreDSLLegalizeCF/CoreDSLLegalizeCF.cpp:138]:
```
  if (isaxOp->walk<ORDER>(RegionToSCFConverter<coredsl::InstructionOp>{*this, changed}).wasInterrupted() ||
      isaxOp->walk<ORDER>(RegionToSCFConverter<coredsl::AlwaysOp>{*this, changed}).wasInterrupted() ||
      isaxOp->walk<ORDER>(RegionToSCFConverter<coredsl::SpawnOp>{*this, changed}).wasInterrupted() ||
      isaxOp->walk<ORDER>(RegionToSCFConverter<func::FuncOp>{*this, changed}).wasInterrupted() ||
      isaxOp->walk<ORDER>(RegionToSCFConverter<scf::ExecuteRegionOp>{*this, changed}).wasInterrupted())
    return signalPassFailure();
```

Comment (7FM) [reply lib/Conversion/CoreDSLLegalizeCF/CoreDSLLegalizeCF.cpp:138]: or some template magic:
```
  auto walkAll = [&]<typename... Ops>() {
    return (isaxOp->walk<ORDER>(RegionToSCFConverter<Ops>{*this, changed})
                .wasInterrupted() ||
            ...);
  };
  if (walkAll.template operator()<coredsl::InstructionOp, coredsl::AlwaysOp,
                                  coredsl::SpawnOp, func::FuncOp,
                                  scf::ExecuteRegionOp>())
    return signalPassFailure();
```

File: lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+struct CoreDSLSwitchToIf
+    : public mlir::shortnail::impl::CoreDSLSwitchToIfBase<CoreDSLSwitchToIf> {
+  using CoreDSLSwitchToIfBase::CoreDSLSwitchToIfBase;
+  // This needs to be a class rather than a lambda because we cannot have
+  // templated lambdas and this needs to be explicitly instantiated with
+  // multiple op types for the walk method
+  template <typename OpType>
+  struct RegionToSCFConverter {
+    // So we can use Pass::getChildAnalysis
+    friend struct CoreDSLSwitchToIf;
```
Comment (7FM) [inline lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp]: is this really needed?

File: test/Conversion/CoreDSLSwitchToIf/switch_test.mlir [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+
+coredsl.isax "SwitchStmt" {
+// CHECK: coredsl.isax "SwitchStmt"
+  coredsl.register core_x @X[32] : ui32
+  coredsl.register local @PC : ui32
+  coredsl.addrspace core_mem @MEM : (ui32) -> ui8
+  coredsl.register local const @MY_CONST = 3 : ui32
+  coredsl.instruction @Simple {lil.enc_immediates = [[["%TREENAIL_WAS_HERE_rs2_4_0", 4, 0, 0, "rs2"]], [["%TREENAIL_WAS_HERE_rs1_4_0", 4, 0, 0, "rs1"]], [["%TREENAIL_WAS_HERE_rd_4_0", 4, 0, 0, "rd"]]]} ("0000000", %TREENAIL_WAS_HERE_rs2_4_0 : ui5, %TREENAIL_WAS_HERE_rs1_4_0 : ui5, "001", %TREENAIL_WAS_HERE_rd_4_0 : ui5, "0101011") {
+
+// CHECK: %0 = hwarith.constant 1 : ui1
```
Comment (7FM) [inline test/Conversion/CoreDSLSwitchToIf/switch_test.mlir]: you should run  your test cases through something like mlir/utils/generate-test-checks.py

**Comment by 7FM** [comment]
Is it possible with treenail to get empty switch statements or one with only a default case? Does the code work with these edge cases?

File: test/Conversion/CoreDSLSwitchToIf/switch_test.mlir [hunk 13]
```diff
@@ -0,0 +1,242 @@
+// RUN: shortnail-opt %s -coredsl-switch-to-if -canonicalize | shortnail-opt | FileCheck %s
+
+coredsl.isax "SwitchStmt" {
+// CHECK: coredsl.isax "SwitchStmt"
+  coredsl.register core_x @X[32] : ui32
+  coredsl.register local @PC : ui32
```
Comment (7FM) [inline test/Conversion/CoreDSLSwitchToIf/switch_test.mlir]: unused?

File: test/Conversion/CoreDSLSwitchToIf/switch_test.mlir [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+  coredsl.addrspace core_mem @MEM : (ui32) -> ui8
+  coredsl.register local const @MY_CONST = 3 : ui32
```
Comment (7FM) [inline test/Conversion/CoreDSLSwitchToIf/switch_test.mlir]: unused?

File: lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp (hunk 11 above)
Comment (PetrifiedPanda) [reply lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp]: It should be needed, because Pass::getChildAnalysis is a protected method, but I tried to delete it and the code still compiles, which is weird, because it gave me a compiler error before. But that was when the CF to SCF lowering was still a separate pass

File: lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp (hunk 8 above)
Comment (PetrifiedPanda) [reply lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp]: Forgot to delete that one. Fixed: https://github.com/esa-tu-darmstadt/Shortnail/pull/10/changes/4b1dfd173a9f4c7996bda023c2d1c6c0e8477bf0

File: lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp (hunk 9 above)
Comment (PetrifiedPanda) [reply lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp]: That works. I didn't know about use_empty(). Also added same thing for hwarith::CastOp, because those are unused as well. https://github.com/esa-tu-darmstadt/Shortnail/pull/10/changes/0d0495fcacb88ca440c45228e41588d3e3be1938

File: test/Conversion/CoreDSLSwitchToIf/switch_test.mlir (hunk 13 above)
Comment (PetrifiedPanda) [reply test/Conversion/CoreDSLSwitchToIf/switch_test.mlir]: Fixed: https://github.com/esa-tu-darmstadt/Shortnail/pull/10/changes/5eb5e8b847256a0f964bf03bb04e8527af32feb7

File: test/Conversion/CoreDSLSwitchToIf/switch_test.mlir (hunk 14 above)
Comment (PetrifiedPanda) [reply test/Conversion/CoreDSLSwitchToIf/switch_test.mlir]: Fixed: https://github.com/esa-tu-darmstadt/Shortnail/pull/10/changes/5eb5e8b847256a0f964bf03bb04e8527af32feb7

File: lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp (hunk 11 above)
Comment (PetrifiedPanda) [reply lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp]: I think if anything, CoreDSLSwitchToIf should declare RegionToSCFConverter as friend, but as it compiles anyways, I just removed the friend declaration: https://github.com/esa-tu-darmstadt/Shortnail/pull/10/changes/5d8cc9952bf3ee85a593cac7ab4ae5acd870758e

File: lib/Conversion/CoreDSLLegalizeCF/CoreDSLLegalizeCF.cpp (hunk 10 above)
Comment (PetrifiedPanda) [reply lib/Conversion/CoreDSLLegalizeCF/CoreDSLLegalizeCF.cpp:138]: Used the first one, because explicit template arguments for lambdas seem to be a C++20 feature: https://github.com/esa-tu-darmstadt/Shortnail/pull/10/changes/3ea3f369060115c8443ae372ed8944932446c742

**Comment by PetrifiedPanda** [comment]
I hadn't thought of that. Pretty sure this won't work in that case. I'll add this to the tests in treenail to see how the generated code looks and then fix it here

**Comment by PetrifiedPanda** [comment]
Switch statements with only a default case work as well, because the canonicalizer seems to remove cf.switch operations with only a default branch. Added a test for this here https://github.com/esa-tu-darmstadt/Shortnail/pull/10/changes/e694af8cdd5c339b92a25cca82cd9127ad5f62ae

File: lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+    if (failed(applyFullConversion(isaxOp, target, std::move(patterns)))) {
+      return signalPassFailure();
+    }
+    // After converting the scf::IndexSwitchOps, the arith::IndexCastOp and
+    // hwarith::CastOps do not have uses anymore. Deleting the
+    // arith::IndexCastOps is especially important, because longnail does not
+    // know how to deal with them
+    // First delete the IndexCastOps, then the hwarith::CastOps, because the
+    // IndexCastOp is their only use
+    isaxOp->walk([](arith::IndexCastOp op) {
```
Comment (7FM) [inline lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp]: is it arith::IndexCastOp or arith::IndexCastUIOp or potentially both?

**Comment by 7FM** [comment]
Wdyt about renaming the pass to smth like CoreDSLLegalizeCF?  the cf -> scf transformation might get relevant in the future for generalized loop support

**Comment by PetrifiedPanda** [comment]
I thought about this as well, but first kept the name, because I thought it could be renamed when other conversions are added. It may be better to rename it anyways so it is easier to see where to add any CF related conversions

File: lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp (hunk 15 above)
Comment (PetrifiedPanda) [reply lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp]: It should only be IndexCastUIOp. But I just realized that commenting out both of the walk operations also does not lead to crashes. Maybe the canonicalizer removes unused operations anyways?

Comment (7FM) [reply lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp]: remove the cleanup code then^^ canonicalize / cse to the rescue

Comment (PetrifiedPanda) [reply lib/Conversion/CoreDSLSwitchToIf/CoreDSLSwitchToIf.cpp]: Done: https://github.com/esa-tu-darmstadt/Shortnail/pull/10/changes/0505ca6f9415db46b345024d310334f29a0832d5

**Review by 7FM — APPROVED** [review]
lgtm, thx!
